src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -32342,9 +32284,312 @@ static void F1<T1>(T1 t1)
+                // (6,43): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                //     static void F3<T>() where T : struct, T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(6, 43),
+                // (7,20): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //     static void F4<T>() where T : A, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(7, 20),
+                // (10,55): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                //     static void F7<T, U>() where T : struct where U : T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(10, 55),
+                // (4,20): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //     static void F1<T>() where T : T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(4, 20));
+        }
+
+        [Fact]
+        public void NullableTInConstraint_05()
+        {
+            var source =
+@"#pragma warning disable 8321
+class C
+{
+    static void M()
+    {
+        void F1<T>() where T : T? { }
+        void F2<T>() where T : class, T? { }
+        void F3<T>() where T : struct, T? { }
+        void F4<T>() where T : A, T? { }
+        void F5<T, U>() where U : T? { }
+        void F6<T, U>() where T : class where U : T? { }
+        void F7<T, U>() where T : struct where U : T? { }
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,17): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //         void F1<T>() where T : T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(6, 17),
+                // (7,17): error CS0454: Circular constraint dependency involving 'T' and 'T'
+                //         void F2<T>() where T : class, T? { }
+                Diagnostic(ErrorCode.ERR_CircularConstraint, "T").WithArguments("T", "T").WithLocation(7, 17),
+                // (8,40): error CS0701: 'T?' is not a valid constraint. A type used as a constraint must be an interface, a non-sealed class or a type parameter.
+                //         void F3<T>() where T : struct, T? { }
+                Diagnostic(ErrorCode.ERR_BadBoundType, "T?").WithArguments("T?").WithLocation(8, 40),
+                // (9,32): error CS0246: The type or namespace name 'A' could not be found (are you missing a using directive or an assembly reference?)
```

Comment:
>The type or namespace name 'A' [](start = 41, length = 30)

Is the error on missing type `A` intentional?

Reply:
No, not intentional.

---
In reply to: [197944764](https://github.com/dotnet/roslyn/pull/27525#discussion_r197944764) [](ancestors = 197944764)